stellargraph/data/explorer.py

```diff
@@ -789,3 +793,136 @@ def _check_neighbourhood_sizes(self, in_size, out_size):
+
+        walks = []
+        for node in nodes:  # iterate over root nodes
+            for walk_number in range(n):  # generate n walks per root node
+                walk = self._walk(
+                    node, length, bidirectional, initial_edge_bias, walk_bias, np_rs
+                )
+                walks.append(walk)
+        return walks
+
+    def _exp_biases(self, times, t_0, decay):
+        # t_0 assumed to be smaller than all time values
+        dist = np.exp(t_0 - np.array(times) if decay else np.array(times) - t_0)
+        sum_dist = np.sum(dist)
+        return dist / sum_dist
+
+    def _biases(self, times, t_0, bias_type, is_forward):
+        if bias_type == "exponential":
+            # exponential decay bias needs to be reversed if looking backwards in time
+            return self._exp_biases(times, t_0, decay=is_forward)
+        else:
+            raise ValueError("Unsupported bias type")
+
+    def _step(self, node, time, is_forward, bias_type, np_rs):
+        """
+        Perform 1 temporal step from a node. Returns None if a dead-end is reached.
+
+        """
+
+        def check_time(t):
+            # time is None indicates no temporal filtering (i.e. initial edge selection)
+            return (
+                (time is None)
+                or (is_forward and t > time)
+                or (not is_forward and t < time)
+            )
+
+        neighbours = [
+            (neighbour, t)
+            for neighbour, t in self.graph.neighbors(node, include_edge_weight=True)
+            if check_time(t)
+        ]
+
+        if len(neighbours) > 0:
+            if bias_type is not None:
+                times = [t for _, t in neighbours]
+
+                # time is None indicates we should obtain the minimum available time for t_0
+                t_0 = time if time is not None else min(times)
+                biases = self._biases(
+                    times, t_0, bias_type=bias_type, is_forward=is_forward
+                )
+            else:
+                # default to uniform random sampling
+                biases = None
+            chosen_index = np_rs.choice(len(neighbours), p=biases)
+            next_node, next_time = neighbours[chosen_index]
+            return next_node, next_time
+        else:
+            return None
+
+    def _walk(self, node, length, bidirectional, initial_edge_bias, walk_bias, np_rs):
+        def step(n, t, is_forward, bias_type=walk_bias):
+            return self._step(
+                n, time=t, is_forward=is_forward, bias_type=bias_type, np_rs=np_rs
+            )
+
+        # sample 1 edge to obtain starting time - any biases are applied as though we are
+        # looking backwards in time.
+        forward = step(node, None, is_forward=False, bias_type=initial_edge_bias)
+        if forward is None:
+            return [node]
+        backward = node, forward[1]
+
+        walk = []
+
+        # take steps until walk is of correct length or reached dead ends
```

Comment:
Per the discussion in section 2.3.3 of http://ryanrossi.com/pubs/nguyen-et-al-WWW18-BigNet.pdf , this is more of a maximum than a "correct" length:

```suggestion
        # take steps until walk is of maximum length or reached dead ends
```

In addition, that section talks about imposing a minimum length on the walks (`ω`); is that something we should do too?

Reply:
I suspect it may be discarded because it's too short to have information from the full context window being used for word2vec (or whatever), so may break things? I don't know.

Reply:
I guess it's easy enough to add, but IMO I don't really see how it applies to temporal walks specifically, rather than random walks in general 🤷‍♂  and I would prefer to avoid adding more parameters that don't get used... but I guess following the paper "correctly" is also important. Do you prefer we add it?